src/nodes.coffee

```diff
@@ -747,13 +747,48 @@ exports.NaNLiteral = class NaNLiteral extends NumberLiteral
     if o.level >= LEVEL_OP then @wrapInParentheses code else code
 
 exports.StringLiteral = class StringLiteral extends Literal
+  constructor: (@originalValue, {@quote, @initialChunk, @finalChunk, @indent, @double} = {}) ->
+    super ''
+    @fromSourceString = @quote?
+    @quote ?= '"'
+    @originalValue = unwrapStringObject @originalValue
+    @formatValue()
+
+  formatValue: ->
+    heredoc = @quote.length is 3
+    @value = do =>
```

Comment:
Why the wrapper of `do =>`?

Reply:
@GeoffreyBooth I'm a fan of this pattern as a way of mini-encapsulating multi-step logic for calculating a single value - by nesting it inside `do =>`, you know that everything inside is geared towards calculating `@value`

You could also extract the value calculation into a named method, but here the logic feels like it belongs inline inside `formatValue()`. Happy to refactor to just assign to `@value` at the end if this pattern feels weird

Reply:
`formatValue` is really just an extension of the constructor, isn’t it? You’d never do something like `stringLiteralNode = new StringLiteral(args); stringLiteralNode.formatValue()`. Since that’s the case, it feels like all this logic belongs in the constructor.

I tend to avoid things like `do =>` because it’s been ingrained into me that creating a new function comes with a performance cost. Maybe that’s not the case realistically anymore, so I’m not sure what to do here. I know that for best performance, we would want to define the function outside of the class itself (or as a separate method in the class) so that it only gets _defined_ once, and executed lots and lots of times for every `StringLiteral` class that gets instantiated; `do =>` both defines _and_ executes it for each new instantiation. But again, I’m not sure how big a hit this is, if it’s worth optimizing for; it’s probably more important to keep logic that deserves to be part of the constructor actually in the constructor, rather than outside of the class entirely.